crates/parser/src/abi/parser.rs

```diff
@@ -71,8 +73,38 @@ impl AbiParser {
             Self::collect_entry_token(entry, &mut token_candidates)?;
         }
 
+        // In theory this is not enough, as after deepening collisions are still possible
+        // Although possible in theory should not be a problem in practice.
+        let mut token_name_occurence: HashMap<String, usize> = HashMap::new();
+        for (_, tokens) in &token_candidates {
+            for token in tokens {
+                let name = token.type_name();
+                let val = token_name_occurence.entry(name).or_insert(0);
+                *val += 1;
+            }
+        }
+
         let tokens = Self::filter_struct_enum_tokens(token_candidates);
 
+        let mut additional_aliases: HashMap<String, String> = HashMap::new();
+        for (_, t) in &tokens {
+            let path = t.type_path();
+            // Crotch to handle cases when type with the same typename are used
+            // in same contract. For example:
+            // enum Event {
+            //   Event1(namespace1::Event),
+            //   Event2(namespace2::Event)
+            // }
+            // When name that occures several times is spotted we register a type alias
+            // (only if there is none). Will apply those later.
```

Comment:
Same here, you should also provide the output of your example.

You could actually refactor this logic into a function (including the one looping on the tokens to get the name occurence count).

Like so, if the proposition of making this depth dynamic sounds good to you, we could avoid this extra loop when the user doesn't want cainome to auto-resolve the conflicts.

You could like so provide the example + the detail of the behavior in the function documentation. 👍